org.eclipse.dsp4j.jsonrpc/src/main/java/org/eclipse/dsp4j/jsonrpc/adapters/DebugMessageTypeAdapterFactory.java

```diff
@@ -0,0 +1,493 @@
+
+	public DebugMessageTypeAdapterFactory(MessageJsonHandler handler) {
+		this.handler = handler;
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> typeToken) {
+		if (!Message.class.isAssignableFrom(typeToken.getRawType()))
+			return null;
+		return (TypeAdapter<T>) new Adapter(handler, gson);
+	}
+
+	private static class Adapter extends TypeAdapter<Message> {
+
+		private final MessageJsonHandler handler;
+		private final Gson gson;
+
+		Adapter(MessageJsonHandler handler, Gson gson) {
+			this.handler = handler;
+			this.gson = gson;
+		}
+
+		@Override
+		public Message read(JsonReader in) throws IOException {
+			if (in.peek() == JsonToken.NULL) {
+				in.nextNull();
+				return null;
+			}
+
+			in.beginObject();
+			String messageType = null, id = null, seq = null, request_seq = null, method = null, message = null;
+			boolean success = true;
+			Object rawParams = null;
+			Object body = null;
+			while (in.hasNext()) {
+				String name = in.nextName();
+				switch (name) {
+				case "seq": {
+					seq = in.nextString();
+					break;
+				}
+				case "request_seq": {
+					// on responses we treat the request_seq as the id
+					request_seq = in.nextString();
+					break;
+				}
+				case "type": {
+					messageType = in.nextString();
+					break;
+				}
+				case "success": {
+					success = in.nextBoolean();
+					break;
+				}
+				case "command": {
+					method = in.nextString();
+					break;
+				}
+				case "event": {
+					method = in.nextString();
+					break;
+				}
+				case "message": {
+					message = in.nextString();
+					break;
+				}
+				case "arguments": {
+					rawParams = parseParams(in, method);
+					break;
+				}
+				case "body": {
+					if ("event".equals(messageType)) {
+						body = parseParams(in, method);
+					} else if ("response".equals(messageType)) {
+						body = parseResult(in, request_seq, success);
+					}
+					break;
+				}
+				default:
+					in.skipValue();
+				}
+			}
+			id = request_seq != null ? request_seq : seq;
+			in.endObject();
+			Object params = parseParams(rawParams, method);
+			return createMessage(messageType, id, method, success, message, params, body);
+		}
+
+		private Object parseResult(JsonReader in, String id, boolean success) {
+			Object result;
+			Type type = null;
+			if (success) {
+				MethodProvider methodProvider = handler.getMethodProvider();
+				if (methodProvider != null && id != null) {
+					String resolvedMethod = methodProvider.resolveMethod(id);
+					if (resolvedMethod != null) {
+						JsonRpcMethod jsonRpcMethod = handler.getJsonRpcMethod(resolvedMethod);
+						if (jsonRpcMethod != null)
+							type = jsonRpcMethod.getReturnType();
+					}
+				}
+			}
+			if (type == null)
+				result = new JsonParser().parse(in);
+			else
+				result = gson.fromJson(in, type);
+			return result;
+		}
+
+		protected Object parseParams(JsonReader in, String method) throws IOException {
```

Comment:
We should make `MessageTypeAdapterFactory` extensible in order to avoid duplicating code here. I'll prepare a PR for that.

Reply:
OK, there is certainly some duplication between the two.

Reply:
Merged #135.

Reply:
I updated the debug code to to use this new code. I also slightly refactored the `MessageTypeAdapterFactory` to access some of its logic from the subclass.